### PR TITLE
Jobs Index and Grid Improvements

### PR DESCRIPTION
**Pull request by jmchilton**
Gets the backend and frontend for Job related stuff prepared to use pagination and other WorkflowList/Invocations improvements downstream.

-  Make the frontend grid more reactive - so the text always matches the query, etc... Matches Galaxy's user experience better from other places in the app.
- Move index logic from the API into the manager layer where it belongs.
- Implement backend search of jobs grid to match Workflows and to allow pagination in a future commit.
- Switch admin Jobs component to use a provider to match the other grids referenced. Add test case for this provider.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/schema/schema.py [hunk 1]
```diff
@@ -1066,6 +1066,28 @@ class WorkflowIndexQueryPayload(Model):
...
 
+class JobIndexSortByEnum(str, Enum):
+    create_time = "create_time"
+    update_time = "update_time"
+
+
+class JobIndexQueryPayload(Model):
+    states: Optional[List[str]] = None
+    user_details: bool = False
+    user_id: Optional[str] = None
```
Comment (mvdbeek) [inline lib/galaxy/schema/schema.py]:
```suggestion
    user_id: Optional[DecodedDatabaseIdField] = None
```

File: lib/galaxy/schema/schema.py [hunk 2]
```diff
@@ -1066,6 +1066,28 @@ class WorkflowIndexQueryPayload(Model):
...
+    tool_ids: Optional[List[str]] = None
+    tool_ids_like: Optional[List[str]] = None
+    date_range_min: Optional[str] = None
+    date_range_max: Optional[str] = None
+    history_id: Optional[str] = None
+    workflow_id: Optional[str] = None
+    invocation_id: Optional[str] = None
```
Comment (mvdbeek) [inline lib/galaxy/schema/schema.py]:
```suggestion
    history_id: Optional[DecodedDatabaseIdField] = None
    workflow_id: Optional[DecodedDatabaseIdField] = None
    invocation_id: Optional[DecodedDatabaseIdField] = None
```

File: lib/galaxy/webapps/galaxy/api/jobs.py [hunk 3]
```diff
@@ -147,12 +142,23 @@ class JobIndexSortByEnum(str, Enum):
     description="Return jobs starting from this specified position. For example, if ``limit`` is set to 100 and ``offset`` to 200, jobs 200-299 will be returned.",
 )
 
+query_tags = [
+    IndexQueryTag("user", "The user email of the user that executed the Job.", "u"),
+    IndexQueryTag("tool_id", "The tool ID corresponding to the job.", "t"),
+    IndexQueryTag("runner", "The job runner name used to execte the job.", "r"),
+    IndexQueryTag("handler", "The job handler name used to execute the job.", "h"),
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/jobs.py]: This doesn't feel like something we should expose to non-admins, those are volatile and at the discretion of the admin. Can we exclude these from the schema ?

File: lib/galaxy/managers/jobs.py [hunk 4]
```diff
@@ -70,6 +84,136 @@ def __init__(self, app: StructuredApp):
...
+        query = build_and_apply_filters(query, payload.tool_ids, lambda t: model.Job.tool_id == t)
+        query = build_and_apply_filters(query, payload.tool_ids_like, lambda t: model.Job.tool_id.like(t))
+        query = build_and_apply_filters(query, payload.date_range_min, lambda dmin: model.Job.update_time >= dmin)
+        query = build_and_apply_filters(query, payload.date_range_max, lambda dmax: model.Job.update_time <= dmax)
+
+        history_id = payload.history_id
+        workflow_id = payload.workflow_id
+        invocation_id = payload.invocation_id
+        if history_id is not None:
+            decoded_history_id = security.decode_id(history_id)
```
Comment (mvdbeek) [inline lib/galaxy/managers/jobs.py]: Can we remove the id decoding from the manager ? These should be integers / DecodedDatabaseIds at this point.

File: lib/galaxy/managers/jobs.py [hunk 5]
```diff
@@ -70,6 +84,136 @@ def __init__(self, app: StructuredApp):
...
         self.dataset_manager = DatasetManager(app)
 
+    def index_query(self, trans, payload: JobIndexQueryPayload):
+        security = trans.security
+        is_admin = trans.user_is_admin
+        user_details = payload.user_details
+
+        user_id = payload.user_id
+        if user_id:
+            decoded_user_id = security.decode_id(user_id)
```
Comment (mvdbeek) [inline lib/galaxy/managers/jobs.py]: This should ideally be decoded already.

**Review by mvdbeek — COMMENTED** [review]
Would be great if we can avoid the id decoding in the manager.

File: lib/galaxy/webapps/galaxy/api/jobs.py (hunk 3 above)
Comment (jmchilton) [reply lib/galaxy/webapps/galaxy/api/jobs.py]: It does only work for admins - but I should document that someway I guess.

**Comment by jmchilton** [comment]
I documented those properties as admin-only.

Screen Shot 2022-05-11 at 11 21 49 AM

Also just annotating everything with the proper encoded/decoded types fixed all the encoding issues because you @mvdbeek are brilliant! Very cool stuff you did with that.

**Comment by bgruening** [comment]
The failing test looks related.

**Comment by jmchilton** [comment]
Ugh... it was all good but I brought in some more commits. I see the problem and will rebase with a fix. Thanks for the catch.

**Pull request merged**